File: intellij-plugin/src/main/kotlin/dev/skillbill/intellij/presentation/SkillBillStatusViewModel.kt
package dev.skillbill.intellij.presentation

import dev.skillbill.intellij.application.StatusRefreshCoordinator
import dev.skillbill.intellij.domain.SkillBillStatusOutcome
import dev.skillbill.intellij.domain.StatusClock
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Project-scoped ViewModel. Exposes immutable [StateFlow] of UI state, maps
 * domain outcomes exhaustively, accepts explicit refresh/lifecycle intents, and
 * derives elapsed durations via an injected clock. Contains no process, JSON,
 * filesystem, or IntelliJ status-bar rendering code.
 */
class SkillBillStatusViewModel(
    private val coordinator: StatusRefreshCoordinator,
    private val clock: StatusClock,
    private val scope: CoroutineScope,
    private val tickIntervalMs: Long = 1_000L,
) {
    private val _uiState = MutableStateFlow<SkillBillStatusUiState>(SkillBillStatusUiState.Idle())
    val uiState: StateFlow<SkillBillStatusUiState> = _uiState.asStateFlow()

    @Volatile
    private var latestOutcome: SkillBillStatusOutcome? = null
    private var tickJob: Job? = null
    private var collectJob: Job? = null
    private var started = false

    fun onWidgetActivated() {
        if (!started) {
            started = true
            collectJob = scope.launch {
                coordinator.outcomes.collect { outcome ->
                    if (outcome != null) {
                        latestOutcome = outcome
                        publish(outcome)
                    }
                }
            }
            tickJob = scope.launch {
                while (isActive) {
                    delay(tickIntervalMs)
                    latestOutcome?.let { publish(it) }
                }
            }
        }
        coordinator.addConsumer()
    }

    fun onWidgetDeactivated() {
        coordinator.removeConsumer()
    }

    fun refresh() {
        coordinator.requestRefresh()
    }

    fun dispose() {
        tickJob?.cancel()
        collectJob?.cancel()
        tickJob = null
        collectJob = null
        started = false
    }

    private fun publish(outcome: SkillBillStatusOutcome) {
        _uiState.value = StatusUiMapper.map(outcome, clock.now())
    }
}

File: vscode-extension/src/presentation/SkillBillStatusViewModel.ts
import { StatusRefreshCoordinator } from "../application/StatusRefreshCoordinator";
import { SkillBillStatusOutcome } from "../domain/SkillBillStatusOutcome";
import { StatusClock } from "../domain/StatusClock";
import { StatusUiMapper } from "./StatusUiMapper";
import { SkillBillStatusUiState } from "./SkillBillStatusUiState";

export type UiStateListener = (state: SkillBillStatusUiState) => void;

export class SkillBillStatusViewModel {
  private latestOutcome: SkillBillStatusOutcome | undefined;
  private tickTimer: ReturnType<typeof setInterval> | undefined;
  private unsubscribe: (() => void) | undefined;
  private started = false;
  private readonly listeners = new Set<UiStateListener>();

  constructor(
    private readonly coordinator: StatusRefreshCoordinator,
    private readonly clock: StatusClock,
    private readonly tickIntervalMs = 1000,
  ) {}

  subscribe(listener: UiStateListener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  onConsumerActivated(): void {
    if (!this.started) {
      this.started = true;
      this.unsubscribe = this.coordinator.subscribe((outcome) => {
        this.latestOutcome = outcome;
        this.publish(outcome);
      });
      this.tickTimer = setInterval(() => {
        if (this.latestOutcome) {
          this.publish(this.latestOutcome);
        }
      }, this.tickIntervalMs);
    }
    this.coordinator.addConsumer();
  }

  onConsumerDeactivated(): void {
    this.coordinator.removeConsumer();
  }

  refresh(): void {
    this.coordinator.requestRefresh();
  }

  dispose(): void {
    if (this.tickTimer) {
      clearInterval(this.tickTimer);
      this.tickTimer = undefined;
    }
    this.unsubscribe?.();
    this.unsubscribe = undefined;
    this.started = false;
    this.listeners.clear();
  }

  private publish(outcome: SkillBillStatusOutcome): void {
    const state = StatusUiMapper.map(outcome, this.clock.now());
    for (const listener of this.listeners) {
      listener(state);
    }
  }
}
